robot_modeling/DQ_SerialManipulator.m

```diff
@@ -289,48 +194,18 @@ function set_effector(obj,effector)
             % This function does not take into account any base or
             % end-effector displacements.
             
-            if nargin == 4
-                n = ith;
-                x_effector = obj.raw_fkm(q,ith);
-                J = obj.raw_pose_jacobian(q,ith);
-                vec_x_effector_dot = J*q_dot(1:ith);
+            if nargin == 4 && ith < obj.n_links
```

Comment:
I'm not sure about this second condition (`ith < obj.n_links`). First, it requires the knowledge of the number of links in an abstract class, which is odd. This should be something evaluated in the concrete subclasses. 

Second, if `ith == obj.n_links`, then the displacement of the base is not considered. Without seeing the implementation of the concrete class, I imagine that you consider there both the displacement of the base and the end-effector. If that is the case, this is not how it's supposed to work. When `ith == obj.n_links`, only the base is considered too.

Reply:
Hi @bvadorno, 

I believe you are referring to the internal implementation of the method `pose_jacobian()` :
```matlab
function J = pose_jacobian(obj, q, ith)
    if nargin == 3 && ith < obj.n_links
        J = hamiplus8(obj.reference_frame)*obj.raw_pose_jacobian(...
            q, ith);
    else
        J = hamiplus8(obj.reference_frame)*haminus8(obj.effector)*...
            obj.raw_pose_jacobian(q);
    end
end
```

- " _Second, if ith == obj.n_links, then the displacement of the base is not considered._  "

I get lost here. If ith == obj.n_links, the `else` condition is executed. In that case, both the base `obj.reference_frame` and the effector `obj.effector` are considered in the current implementation of the PR. (Please correct me if I am missing something).

- " _Without seeing the implementation of the concrete class, I imagine that you consider there both the displacement of the base and the end-effector. If that is the case, this is not how it's supposed to work._ "

Yes. All raw_methods ( `raw_fkm`, `raw_pose_jacobian`, `raw_pose_jacobian_derivative`) implemented in the concrete class `DQ_SerialManipulatorDH` are supposed to ignore both the base and the effector. If this is not the case, I need to fix it.
This is the implementation of  `raw_pose_jacobian`  (I removed the comments here)
```matlab
function J = raw_pose_jacobian(obj,q,to_ith_link)           
            if nargin < 3
                to_ith_link = obj.n_links;
            end
            x_effector = obj.raw_fkm(q,to_ith_link);
            
            x = DQ(1);
            J = zeros(8,to_ith_link);
            
            for i = 0:to_ith_link-1
                w = obj.get_w(i+1);
                z = 0.5*Ad(x,w);
                x = x*obj.dh2dq(q(i+1),i+1);
                j = z * x_effector;
                J(:,i+1) = vec8(j);
            end
end
```

- When ith == obj.n_links, only the base is considered too.

Assuming that I'm not missing something,  If ith == obj.n_links, the `else` condition of `DQ_SerialManipulator. pose_jacobian` is executed, and therefore, both the base `obj.reference_frame` and the effector `obj.effector` are considered. 

You mentioned in other comments of this PR that if `ith == obj.n_links`, the end-effector transformation is not supposed to be considered. 

I need to fix it to accomplish the expected behavior.

Reply:
Hi @juanjqo,

Yes, you are right. As I mentioned before I don't remember why I changed the initial specification that I had envisaged for the `pose_jacobian()`. Also, I missed this part of the conversation because the review process becomes chaotic very quickly. In any case, we settled the matter. Hence, I'm resolving this conversation.

Thanks,
Bruno